src/hotspot/share/prims/jni.cpp

```diff
@@ -963,7 +963,9 @@ JNI_ENTRY(jobject, jni_AllocObject(JNIEnv *env, jclass clazz))
   jobject ret = nullptr;
   DT_RETURN_MARK(AllocObject, jobject, (const jobject&)ret);
 
-  instanceOop i = InstanceKlass::allocate_instance(JNIHandles::resolve_non_null(clazz), CHECK_NULL);
+  instanceOop i = InstanceKlass::allocate_instance(
+                    InstanceKlass::cast(java_lang_Class::as_Klass(JNIHandles::resolve_non_null(clazz))),
```

Comment:
Perhaps it would be nice with a utility method to reduce some of the clutter. Just folding the `as_Klass` into a new method `instanceKlass::cast_from_oop(..)` (or just `from_oop`) would cut away a fair chunk and could be used to similar effect elsewhere (counting 21 cases where this would be applicable).

Reply:
An specific RFE is created for this parameter change. The change is reverted and not in this PR anymore.